### PR TITLE
feat(training): low-VRAM smoke config for 2b SFT on 12GB consumer GPUs

### PR DESCRIPTION
**Pull request by 2-A-M**
## Summary

Adds a `--low-vram-smoke` CLI preset to `train_local.py` that brings the
2B SFT path inside a 12 GB envelope so train→quant→bench plumbing can be
validated end-to-end on commodity hardware (RTX 3060 / 4070 class) before
reaching for a rented H100/H200.

Motivation: the registry default for `qwen3.5-2b` targets a 16 GB local
GPU (`seq_len=8192`, `budget=15.5 GB`). On a 12 GB consumer card that
OOMs at the fp32 logits transient (`B*S*V*4` bytes; Qwen vocab=248k makes
this dominant). 12 GB cards are the realistic baseline for the eliza-1
"test + fine-tune" track on individual contributor machines, so the
default should be reachable from them.

## What the preset does

Overrides registry defaults to:

- `seq_len = 2048`                    (registry default 4096 / 8192)
- `per_device_batch_size = 1`
- `gradient_accumulation_steps = 16`  (effective batch stays at 16)
- `max_samples = 1000`
- `epochs = 1`
- `memory_budget_gb = 11.5`           (1.5 GB headroom under 12 GB)

Liger fused chunked-CE and activation checkpointing stay at their
registry / `train_local.py` defaults (both on). APOLLO-Mini (registry
default for the local tiers) remains the optimizer.

The override block snapshots the argparse-default state for each
relevant arg **before** the registry merge, so it can tell the difference
between "value came from the registry" (override OK) and "value came
from the CLI" (do not override). CLI flags the operator passed
explicitly always win over both the registry merge and the preset.

## Files

- `packages/training/scripts/train_local.py` — `--low-vram-smoke` flag +
  override block (`+64` lines)
- `packages/training/scripts/test_train_local_low_vram_smoke.py` —
  6 CPU-only unit tests pinning the override semantics (registry
  interaction, CLI override precedence, no-registry-key path, regression
  guard against removing the flag without removing the block) (`+187`
  lines, all 6 pass on `python3.11 / pytest 9.0.3`)
- `packages/training/README.md` — new "Running on a 12 GB consumer GPU"
  section with the invocation, kernel prerequisites, measured numbers,
  trade-offs, and the OOM-fallback ladder (`+100` net)

## Measured on RTX 3060 (12 GB)

Hardware: NVIDIA GeForce RTX 3060 12 GB, WSL2 Ubuntu, torch
2.12.0+cu130, transformers 5.8.1, accelerate 1.13, trl 1.4, apollo-torch
1.0.4, liger-kernel installed but Triton cannot JIT (no `gcc`,
`python3.x-dev`, or CUDA toolkit in this WSL host → train_local.py logs
the documented fallback and uses HF defaults). 314 records in
`data/final-eliza1-smoke/train.jsonl`.

| step                                              | result            |
|---------------------------------------------------|-------------------|
| `--preflight-only` with `qwen3.5-2b` + preset     | ok (314/314 accepted, optimizer=apollo_mini rank=1, budget=11.5 GB) |
| `--preflight-only` with `qwen3.5-0.8b` + preset   | ok (same shape, smaller geometry) |
| Full SFT, 2B, 128 samples, no Liger / no FLA      | model loaded + optimizer built; first backward saturates VRAM at ~12 GB and does not complete in a useful time window |
| Full SFT, 0.8B, 128 samples, no Liger / no FLA    | model loaded + optimizer built; peak VRAM ~12.0 GB during first backward; step 0/8 runs but each step is dominated by the pure-PyTorch linear-attn fallback |

The instrumentation callback (enabled because `--memory-budget-gb` is
set) fails the run loud the moment `torch.cuda.max_memory_reserved()`
exceeds the budget by more than 10 %.

## Honest caveats — Liger + flash-linear-attention are the gate

Without the Triton kernel toolchain, the 2B-on-12 GB path is effectively
preflight + small-step verification only. Two kernels matter:

- **Liger fused chunked-CE** — cuts the fp32 logits transient ~4–8× on
  Qwen's 248k vocab. Triton JIT-compiles a small CUDA helper at the
  first kernel launch; missing `gcc` / `python3.x-dev` / CUDA toolkit
  → `train_local.py` warns and falls back to HF defaults.
- **flash-linear-attention** — the Qwen3.5 hybrid linear-attn layers
  (6 of 24 in 0.8B/2B) have a fast Triton path via
  https://github.com/fla-org/flash-linear-attention. Without it,
  transformers logs `The fast path is not available because one of the
  required library is not installed. Falling back to torch
  implementation` and each step gets ~10× slower.

The README documents this explicitly so users on stock WSL2 / Windows
hosts don't burn an evening wondering why the smoke is at step 0/8
after 15 minutes. The recommended invocation on those hosts is
`--registry-key qwen3.5-0.8b --low-vram-smoke` — same preset, half the
parameters, comfortably inside 12 GB without Triton.

## Test plan

- [x] `pytest packages/training/scripts/test_train_local_low_vram_smoke.py -v`
  → 6/6 pass on python 3.11 + pytest 9.0.3.
- [x] `pytest packages/training/scripts/training/test_model_registry.py -v`
  → 17/17 pass (no regression in registry tests).
- [x] `python scripts/train_local.py --registry-key qwen3.5-2b --low-vram-smoke --preflight-only`
  → preset logs the override line, registry merge runs first, preflight
  confirms 314/314 records accepted at the new budget.
- [x] `python scripts/train_local.py --registry-key qwen3.5-0.8b --low-vram-smoke --preflight-only`
  → same result against the 0.8B tier.
- [x] End-to-end SFT smoke on RTX 3060: model loads at the new budget,
  first backward saturates VRAM in line with the 11.5 GB budget. Full
  step throughput is gated on Liger / flash-linear-attention being live
  (see the kernel prerequisites note above); the preset is functioning
  as designed.

## Why this is a `feat` not a `chore`

It is a new user-facing CLI flag that changes how `train_local.py`
behaves and lands measured numbers in the public README. Down-tier
behavior is unchanged when the flag is not passed.

Cross-references: shaw's task list item #1 (test eliza-1) and #2
(fine-tune eliza-1) — this lowers the floor on item #2 from a rented
GPU to a $300 consumer card.

<h3>Greptile Summary</h3>

This PR adds a `--low-vram-smoke` CLI preset to `train_local.py` that brings the 2B SFT path inside a 12 GB GPU envelope, and refactors arg parsing into two public helpers (`build_parser()` / `apply_resolved_defaults()`) that use `None`-sentinel defaults to unambiguously distinguish user-supplied CLI values from argparse fill-ins. The README gains a consumer-GPU section with measured numbers, OOM-fallback guidance, and kernel prerequisites.

- **`train_local.py`**: Extracts `build_parser()` and `apply_resolved_defaults()` from `main()`. Tracked args now default to `None`; the merge pipeline is registry → preset → `_FALLBACK_DEFAULTS`. The preset fires only after the registry merge, and only touches fields the user did not explicitly pass.
- **`test_train_local_low_vram_smoke.py`**: Eight new CPU-only tests, including a parametrized table covering all five preset-governed args with values that equal the historical argparse defaults — directly addressing the previous sentinel-default issue flagged on this PR.
- **`README.md`**: Adds a "Running on a 12 GB consumer GPU" section with invocation examples, kernel prerequisites, measured RTX 3060 numbers, trade-offs, and an OOM-fallback ladder.

<h3>Confidence Score: 5/5</h3>

Safe to merge. The change is additive (new flag, new helper functions, new documentation), and the refactoring of the parse/merge pipeline is backward-compatible with all existing invocations.

The three issues surfaced in the previous review round — duplicate merge logic in tests, incomplete CLI-wins coverage for preset-governed args, and the sentinel-default ambiguity — are all addressed cleanly. The None-sentinel pattern unambiguously distinguishes user-supplied values from argparse fill-ins, the factored-out helpers mean tests exercise the exact production code path, and the parametrized test table covers every preset-governed arg at the historically tricky default-equal-to-explicit-value edge case. Downstream consumers of args that previously had real default values are protected by _FALLBACK_DEFAULTS, which fires after all merges, preserving the original runtime behavior when the flag is not set.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/training/scripts/train_local.py | Refactors parser + merge logic into build_parser() / apply_resolved_defaults(); adds --low-vram-smoke preset with correct None-sentinel user_passed detection and three-stage merge (registry → preset → fallback). Logic is sound; all tracked args are non-None before the log.info format strings that reference them. |
| packages/training/scripts/test_train_local_low_vram_smoke.py | Eight CPU-only tests exercising the preset's override semantics. _resolve() delegates to the real build_parser() + apply_resolved_defaults() (no hand-maintained mirror), and the parametrized table covers all previously uncovered preset-governed args at their historical default values. |
| packages/training/README.md | Adds a 100-line 'Running on a 12 GB consumer GPU' section with invocation examples, kernel prerequisites, measured RTX 3060 numbers, trade-off table, and OOM-fallback ladder. Content is accurate and well-structured. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["build_parser().parse_args()"] --> B["apply_resolved_defaults(args)"]
    B --> C{"args.registry_key set?"}
    C -- Yes --> D["Load registry entry\nFill None fields:\nmodel, batch_size, grad_accum,\nmax_seq_len, optimizer,\napollo_rank, memory_budget_gb"]
    C -- No --> E{"args.low_vram_smoke?"}
    D --> E
    E -- Yes --> F["Preset fills remaining None fields\nmax_seq_len=2048, batch=1,\ngrad_accum=16, max_samples=1000,\nepochs=1.0, budget=11.5GB"]
    E -- No --> G["_FALLBACK_DEFAULTS fills\nremaining None fields\nbatch=4, grad_accum=8,\nmax_seq_len=4096, epochs=3.0,\nmax_samples=0, optimizer=apollo"]
    F --> G
    G --> H["main() continues\nwith fully resolved args"]

    style D fill:#d4edda
    style F fill:#fff3cd
    style G fill:#d1ecf1
```

<sub>Reviews (3): Last reviewed commit: ["refactor(training): extract build\_parser..."](https://github.com/elizaos/eliza/commit/7d835045601e2fee3b4fff5c0df1dd743174954c) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=32806902)</sub>

### CODE REVIEW

**Review by lalalune — APPROVED** [review]
LGTM. Clean additive change — the None-sentinel pattern correctly distinguishes user-supplied values from argparse defaults, the three-stage merge (registry → preset → fallback) is sound, and all prior review issues are addressed. 8/8 tests pass.
